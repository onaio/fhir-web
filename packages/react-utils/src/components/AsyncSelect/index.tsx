export * from './BaseAsyncSelect';
export * from './PaginatedAsyncSelect';
export * from './ValueSetAsyncSelect';
