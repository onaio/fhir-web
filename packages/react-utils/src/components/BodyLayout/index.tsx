import React, { ReactNode } from 'react';
import { RichPageHeader, RichPageHeaderProps } from '../PageHeaders/RichPageHeader';

export interface BodyLayoutProps {
  headerProps: RichPageHeaderProps;
  children: ReactNode;
  id?: string;
}

export const BodyLayout = (props: BodyLayoutProps) => {
  const { headerProps, children } = props;
  return (
    <div className="details-full-view">
      <RichPageHeader {...headerProps} />
      <div className="content-body">{children}</div>
    </div>
  );
};
