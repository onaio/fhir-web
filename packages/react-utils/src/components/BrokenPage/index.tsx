/** use this component as the temporary background view ,
 * when the app runs into an unrecoverable error.
 */
import React, { useState } from 'react';
import { Result } from 'antd';
import { sendErrorNotification } from '@opensrp/notifications';
import { ExtraLinks, UtilPageExtraProps, extraLinksDefault } from '../UtilPageExtra';
import { ERROR, SOMETHING_WENT_WRONG } from '../../lang';

/** props for the broken page component */
export interface BrokenPageProps extends UtilPageExtraProps {
  title: string;
  errorMessage?: string;
}

export const defaultProps = {
  ...extraLinksDefault,
  errorMessage: SOMETHING_WENT_WRONG,
  title: ERROR,
};

/** the component that renders a 500 view */

const BrokenPage = (props: BrokenPageProps) => {
  const { errorMessage, title } = props;
  const extraLinksProps = {
    homeUrl: props.homeUrl,
  };
  return (
    <Result
      status="500"
      title={title}
      subTitle={errorMessage}
      extra={<ExtraLinks {...extraLinksProps} />}
    />
  );
};

BrokenPage.defaultProps = defaultProps;

export { BrokenPage };

/** custom hook that abstracts behavior of a broken page */

export const useHandleBrokenPage = () => {
  const [broken, setBroken] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string | undefined>();

  /**
   * Convenience function to handle cases where we must abort and tell the user we have done so
   *
   * @param {Error} error - Error object
   */
  function handleBrokenPage(error: Error) {
    sendErrorNotification(error.name, error.message);
    setErrorMessage(error.message);
    setBroken(true);
  }

  return { broken, errorMessage, handleBrokenPage };
};
