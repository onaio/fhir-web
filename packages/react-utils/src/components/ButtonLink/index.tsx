import React, { HTMLAttributes } from 'react';
import { Button } from 'antd';
import { useHistory } from 'react-router';

/***
 * Simple button, type link supposed to be used for navigation
 */

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  name: string;
  route: string;
}

const ButtonLink = (props: ButtonProps) => {
  const history = useHistory();
  const { name, route, className } = props;

  const handleClick = () => {
    history.push(route);
  };

  return (
    <Button onClick={handleClick} className={className} type="link">
      {name}
    </Button>
  );
};

export { ButtonLink };
