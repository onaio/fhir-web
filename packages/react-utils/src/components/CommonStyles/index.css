/* common styles goes here */
.mr-0 {
  margin-right: 0px;
}
.pt-0 {
  padding-top: 0px !important;
}
.d-flex {
  display: flex;
}
.flex-right {
  display: flex !important;
  justify-content: flex-end !important;
}
