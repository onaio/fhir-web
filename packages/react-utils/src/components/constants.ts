export const HOME_URL = '/';
