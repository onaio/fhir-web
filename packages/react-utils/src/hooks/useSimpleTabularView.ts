import { ChangeEvent, useCallback } from 'react';
import { FHIRServiceClass } from '../helpers/dataLoaders';
import { getQueryParams } from '../components/Search/utils';
import { getResourcesFromBundle } from '../helpers/utils';
import { useQuery } from 'react-query';
import { getConfig } from '@opensrp/pkg-config';
import { useHistory, useLocation, useRouteMatch } from 'react-router';
import type { IBundle } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IBundle';
import { Resource } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/resource';
import { URLParams } from '@opensrp/server-service';
import {
  getNextUrlOnSearch,
  getNumberParam,
  getStringParam,
  pageQuery,
  pageSizeQuery,
  searchQuery,
  startingPage,
  startingPageSize,
} from './utils';

export interface FhirApiFilter {
  page: number;
  pageSize: number;
  search: string | null;
}

export type ExtraParams = URLParams | ((search: string | null) => URLParams);

const defaultGetExtraParams = (search: string | null) => {
  if (search) {
    return { 'name:contains': search };
  }
  return {};
};

/**
 * Unified function that gets a list of FHIR resources from a FHIR hapi server
 *
 * @param baseUrl - base url
 * @param resourceType - resource type as endpoint
 * @param params - our params
 * @param extraParams - any extra user-defined params
 */
const loadResources = async (
  baseUrl: string,
  resourceType: string,
  params: FhirApiFilter,
  extraParams: ExtraParams
) => {
  const { page, pageSize, search } = params;
  let filterParams: URLParams = {};

  let otherParams = extraParams;
  if (typeof extraParams === 'function') {
    otherParams = extraParams(search);
  }

  filterParams = {
    _total: 'accurate',
    ...filterParams,
    ...otherParams,
    _getpagesoffset: (page - 1) * pageSize,
    _count: pageSize,
  };
  const service = new FHIRServiceClass<IBundle>(baseUrl, resourceType);
  const res = await service.list(filterParams);
  if (res.total === undefined) {
    // patient endpoint does not include total after _search response like other resource endpoints do
    const countFilter = {
      ...filterParams,
      _summary: 'count',
    };
    const { total } = await service.list(countFilter);
    res.total = total;
    return res;
  }
  return res;
};

/**
 * Re-usable hook that abstracts search and table pagination for usual list view component
 * Should only be used when getting data from a server that follows the hapi FHIR spec
 *
 * @param fhirBaseUrl - fhir server baser url
 * @param resourceType - resource type as endpoint
 * @param extraParams - further custom search param filters during api requests
 */
export function useSimpleTabularView<T extends Resource>(
  fhirBaseUrl: string,
  resourceType: string,
  extraParams: URLParams | ((search: string | null) => URLParams) = defaultGetExtraParams
) {
  const location = useLocation();
  const history = useHistory();
  const match = useRouteMatch();

  const page = getNumberParam(location, pageQuery, startingPage) as number;
  const search = getStringParam(location, searchQuery);
  const defaultPageSize =
    (getConfig('defaultTablesPageSize') as number | undefined) ?? startingPageSize;
  const pageSize = getNumberParam(location, pageSizeQuery, defaultPageSize) as number;

  type TRQuery = [string, number, number, string, URLParams];
  type QueryKeyType = { queryKey: TRQuery };

  const queryFn = useCallback(
    async ({ queryKey: [_, page, pageSize, search, extraParams] }: QueryKeyType) => {
      return loadResources(fhirBaseUrl, resourceType, { page, pageSize, search }, extraParams);
    },
    [fhirBaseUrl, resourceType]
  );

  const rQuery = {
    queryKey: [resourceType, page, pageSize, search, extraParams] as TRQuery,
    queryFn,
    select: (data: IBundle) => ({
      records: getResourcesFromBundle<T>(data),
      total: data.total ?? 0,
    }),
    keepPreviousData: true,
    staleTime: 5000,
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
  };
  const { data, ...restQueryValues } = useQuery(rQuery);

  const searchFormProps = {
    defaultValue: getQueryParams(location)[searchQuery],
    onChangeHandler: function onChangeHandler(event: ChangeEvent<HTMLInputElement>) {
      const nextUrl = getNextUrlOnSearch(event, location, match);
      history.push(nextUrl);
    },
  };

  const tablePaginationProps = {
    current: page,
    pageSize,
    total: data?.total,
    defaultPageSize,
    onChange: (current: number, pageSize?: number) => {
      if (current && pageSize) {
        const newSParams = new URLSearchParams(location.search);
        newSParams.set(pageSizeQuery, pageSize.toString());
        newSParams.set(pageQuery, current.toString());
        history.push(`${match.url}?${newSParams.toString()}`);
      }
    },
  };

  return {
    tablePaginationProps,
    queryValues: {
      data,
      ...restQueryValues,
    },
    searchFormProps,
  };
}
