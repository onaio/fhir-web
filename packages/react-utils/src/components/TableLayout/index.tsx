import React, { useState } from 'react';
import { Table as AntTable } from 'antd';
import { ColumnType, TableProps as AntTableProps } from 'antd/lib/table';
import { Dictionary } from '@onaio/utils';
import { TABLE_PAGE_SIZE, TABLE_PAGE_SIZE_OPTIONS, TABLE_ACTIONS_KEY } from '../../constants';
import { getConfig, TableState, setConfig } from '@opensrp/pkg-config';
import { Optional } from '../../helpers/utils';
import lang from '../../lang';

// Options Must be of any Data type as the Data could be any
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Options<T = any> = AntTableProps<T>;
type Action<T> = Optional<Column<T>, 'key' | 'dataIndex'>;
type TKey<T> = keyof T & React.Key;

export interface Column<T> extends ColumnType<T>, Dictionary {
  dataIndex?: TKey<T>;
  key?: TKey<T>;
}

export const defaults: Options = {
  pagination: {
    showQuickJumper: true,
    showSizeChanger: true,
    defaultPageSize: TABLE_PAGE_SIZE,
    pageSizeOptions: TABLE_PAGE_SIZE_OPTIONS,
  },
};

interface Props<T> extends Omit<Options<T>, 'columns' | 'dataSource'> {
  datasource: T[];
  columns?: Column<T>[];
  actions?: Action<T>;
}

interface PersistState {
  id: string;
  persistState: boolean;
}

interface NoPersistState {
  id?: string;
  persistState?: never;
}

export type TableProps<T> = Props<T> & (PersistState | NoPersistState);

/** Table Layout Component used to render the table with default Settings
 *
 * @param props - Table settings
 * @returns - the component
 */
export function TableLayout<T extends object = Dictionary>(props: TableProps<T>) {
  const { id, columns, datasource, children, persistState, actions, ...restprops } = props;

  const options: Options = { ...defaults, ...restprops };
  const tablesState = getConfig('tablespref') ?? {};

  if (columns && actions) {
    const actionsColumn: Column<T> = {
      key: TABLE_ACTIONS_KEY as TKey<T>,
      title: lang.ACTIONS,
      ...actions,
    };
    columns.push(actionsColumn);
  }

  const [tableState, setTableState] = useState<TableState>(
    id && tablesState[id] !== undefined ? tablesState[id] : {}
  );

  const tableprops: Options = {
    ...options,
    ...tablesState,
    // bind change event to pagination only if  table is uniquely identifable /and have to persistState
    // instead of table onchange bind it to pagination onchange so that table change event will be free to use for user at will

    // only add pagination event listner if we have id and persistState true
    ...(id &&
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      persistState && {
        pagination: {
          ...options.pagination,
          ...tableState?.pagination,
          onChange: paginationChange,
        },
      }),
  };

  /** Table Layout Component used to render the table with default Settings
   *
   * @param page - the current viewing Page number
   * @param pageSize - the no of rows to show in the table
   */
  function paginationChange(page: number, pageSize?: number) {
    // if table is uniquely identifable and have to persistState then Save Pagination to pkg-config store
    const newstate: TableState = {
      ...tableState,
      pagination: { ...tableState, current: page, pageSize: pageSize },
    };
    tablesState[id as string] = newstate;
    setConfig('tablespref', { ...tablesState });
    setTableState(newstate);
  }

  return (
    <AntTable<T> dataSource={datasource} columns={columns} {...tableprops}>
      {children}
    </AntTable>
  );
}

export default TableLayout;
