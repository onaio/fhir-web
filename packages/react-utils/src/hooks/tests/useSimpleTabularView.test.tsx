import React from 'react';
import { store } from '@opensrp/store';
import { authenticateUser } from '@onaio/session-reducer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { QueryClientProvider, QueryClient } from 'react-query';
import { Router, Route, Switch } from 'react-router';
import { TableLayout } from '../../components/TableLayout';
import { useSimpleTabularView } from '../useSimpleTabularView';
import nock from 'nock';
import { dataPage1, dataPage2, searchData } from './fixtures';
import userEvents from '@testing-library/user-event';
import { Input } from 'antd';

jest.mock('fhirclient', () => {
  return jest.requireActual('fhirclient/lib/entry/browser');
});

const rQClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      cacheTime: 0,
    },
  },
});

// TODO - boiler plate
store.dispatch(
  authenticateUser(
    true,
    {
      email: 'bob@example.com',
      name: 'Bobbie',
      username: 'RobertBaratheon',
    },
    { api_token: 'hunter2', oAuth2Data: { access_token: 'bamboocha', state: 'abcde' } }
  )
);

// we first setup the wrapper components, somewhere to run the hooks during tests
const options = {
  baseUrl: 'http://example.com',
  endpoint: 'data',
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const SearchForm = (props: any) => {
  const { onChangeHandler, ...otherProps } = props;

  return (
    <div className="search-input-wrapper">
      <Input onChange={onChangeHandler} {...otherProps}></Input>
    </div>
  );
};

// minimal app to wrap our hook.
const SampleApp = () => {
  const { tablePaginationProps, queryValues, searchFormProps } = useSimpleTabularView(
    options.baseUrl,
    options.endpoint
  );

  const { data, isFetching, isLoading } = queryValues;

  const columns = [
    {
      title: 'Name/Id',
      dataIndex: 'title',
      width: '20%',
    },
  ];

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const tableProps: any = {
    datasource: data?.records ?? [],
    columns,
    loading: isFetching || isLoading,
    pagination: tablePaginationProps,
  };

  return (
    <div>
      <SearchForm {...searchFormProps} data-testid="search-form" />
      <TableLayout {...tableProps} />
    </div>
  );
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const App = (props: any) => {
  return (
    <Switch>
      <Route exact path="/qr">
        <QueryClientProvider client={rQClient}>{props.children}</QueryClientProvider>
      </Route>
    </Switch>
  );
};

// we now setup the tests
beforeAll(() => {
  nock.disableNetConnect();
});

afterAll(() => {
  nock.enableNetConnect();
});

afterEach(() => {
  nock.cleanAll();
  cleanup();
});

test('pagination and search work correctly', async () => {
  const history = createMemoryHistory();
  history.push('/qr');

  nock(options.baseUrl)
    .get(`/${options.endpoint}/_search`)
    .query({
      _total: 'accurate',
      _getpagesoffset: 0,
      _count: 20,
    })
    .reply(200, dataPage1)
    .persist();

  nock(options.baseUrl)
    .get(`/${options.endpoint}/_search`)
    .query({
      _total: 'accurate',
      _getpagesoffset: 20,
      _count: 20,
    })
    .reply(200, dataPage2)
    .persist();

  nock(options.baseUrl)
    .get(`/${options.endpoint}/_search`)
    .query({
      _total: 'accurate',
      _getpagesoffset: 0,
      _count: 20,
      'name:contains': '345',
    })
    .reply(200, searchData)
    .persist();

  render(
    <Router history={history}>
      <App>
        <SampleApp />
      </App>
    </Router>
  );

  await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

  const waitForSpinner = async () => {
    return await waitFor(() => {
      expect(document.querySelector('.ant-spin')).toBeInTheDocument();
    });
  };

  await waitFor(() => {
    expect(screen.getByText(/NSW Government My Personal Health Record/)).toBeInTheDocument();
  });

  document.querySelectorAll('tr').forEach((tr, idx) => {
    tr.querySelectorAll('td').forEach((td) => {
      expect(td).toMatchSnapshot(`table row ${idx} page 1`);
    });
  });

  fireEvent.click(screen.getByTitle('2'));

  expect(history.location.search).toEqual('?pageSize=20&page=2');

  await waitForSpinner();
  await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

  expect(screen.getByText(/426 - title/)).toBeInTheDocument();
  document.querySelectorAll('tr').forEach((tr, idx) => {
    tr.querySelectorAll('td').forEach((td) => {
      expect(td).toMatchSnapshot(`table row ${idx} page 2`);
    });
  });

  // works with search as well.
  const searchForm = document.querySelector('[data-testid="search-form"]') as Element;
  userEvents.type(searchForm, '345');

  expect(history.location.search).toEqual('?pageSize=20&page=1&search=345');

  await waitForSpinner();
  await waitForElementToBeRemoved(document.querySelector('.ant-spin'));

  document.querySelectorAll('tr').forEach((tr, idx) => {
    tr.querySelectorAll('td').forEach((td) => {
      expect(td).toMatchSnapshot(`Search ${idx} page 1`);
    });
  });

  // remove search.
  userEvents.clear(searchForm);
  expect(history.location.search).toEqual('?pageSize=20&page=1');

  expect(nock.pendingMocks()).toEqual([]);
});
