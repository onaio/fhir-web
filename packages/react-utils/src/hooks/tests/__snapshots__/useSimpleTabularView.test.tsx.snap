// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination and search work correctly: Search 1 page 1 1`] = `
<td
  class="ant-table-cell"
>
  Birth Notification CRVS sample
</td>
`;

exports[`pagination and search work correctly: table row 1 page 1 1`] = `
<td
  class="ant-table-cell"
>
  NSW Government My Personal Health Record
</td>
`;

exports[`pagination and search work correctly: table row 1 page 2 1`] = `
<td
  class="ant-table-cell"
>
  426 - title
</td>
`;

exports[`pagination and search work correctly: table row 2 page 1 1`] = `
<td
  class="ant-table-cell"
>
  219 - title
</td>
`;

exports[`pagination and search work correctly: table row 2 page 2 1`] = `
<td
  class="ant-table-cell"
>
  NSW Government My Personal Health Record
</td>
`;
